File: mvp/webapp/src/store/fundStore.ts
import { create } from 'zustand';
import type { FundInfo, FundNAV, FundHolding, FundDiagnosis } from '@/types';
import { searchFunds as apiSearchFunds, getFundInfo as apiGetFundInfo, getFundNAV as apiGetFundNAV, getFundHoldings as apiGetFundHoldings } from '@/services/fundApi';

interface FundState {
  searchResults: FundInfo[];
  selectedFund: FundDiagnosis | null;
  isLoading: boolean;
  error: string | null;
  navHistory: Map<string, FundNAV[]>;
  holdingsCache: Map<string, FundHolding[]>;

  // Actions
  searchFunds: (keyword: string) => Promise<void>;
  selectFund: (code: string) => Promise<void>;
  getFundInfo: (code: string) => Promise<FundInfo | null>;
  getFundNAV: (code: string) => Promise<FundNAV[]>;
  getFundHoldings: (code: string) => Promise<FundHolding[]>;
  clearError: () => void;
  clearSelection: () => void;
}

export const useFundStore = create<FundState>((set, get) => ({
  searchResults: [],
  selectedFund: null,
  isLoading: false,
  error: null,
  navHistory: new Map(),
  holdingsCache: new Map(),

  searchFunds: async (keyword: string) => {
    set({ isLoading: true, error: null });

    try {
      const results = await apiSearchFunds(keyword);
      set({ searchResults: results, isLoading: false });

      if (results.length === 0) {
        set({ error: '未找到相关基金，请尝试其他关键词' });
      }
    } catch {
      set({ error: '搜索失败，请稍后重试', isLoading: false });
    }
  },

  selectFund: async (code: string) => {
    set({ isLoading: true, error: null });

    try {
      const info = await get().getFundInfo(code);
      if (!info) {
        set({ error: '未找到该基金', isLoading: false });
        return;
      }

      const [navHistory, holdings] = await Promise.all([
        get().getFundNAV(code),
        get().getFundHoldings(code),
      ]);

      // Calculate risk metrics based on real data
      const volatility = calculateVolatility(navHistory);
      const maxDrawdown = calculateMaxDrawdown(navHistory);
      const sharpeRatio = calculateSharpeRatio(navHistory);

      // Calculate industry distribution from holdings
      const industryDistribution: Record<string, number> = {};
      holdings.forEach((h) => {
        if (h.industry) {
          industryDistribution[h.industry] = (industryDistribution[h.industry] ?? 0) + h.proportion;
        }
      });

      // Calculate rating (1-5 stars)
      const rating = Math.round(
        (5 - Math.abs(volatility - 20) / 5) * (1 + sharpeRatio / 2)
      );

      const diagnosis: FundDiagnosis = {
        fundInfo: info,
        navHistory,
        topHoldings: holdings.slice(0, 10),
        industryDistribution,
        riskMetrics: {
          volatility,
          maxDrawdown,
          sharpeRatio,
          correlationToMarket: null,
        },
        rating: Math.max(1, Math.min(5, rating)),
        recommendation: sharpeRatio > 1.2 ? '推荐持有' : '建议观望',
      };

      set({ selectedFund: diagnosis, isLoading: false });
    } catch {
      set({ error: '获取基金信息失败', isLoading: false });
    }
  },

  getFundInfo: async (code: string) => {
    return apiGetFundInfo(code);
  },

  getFundNAV: async (code: string) => {
    const cached = get().navHistory.get(code);
    if (cached) return cached;

    const history = await apiGetFundNAV(code);

    set((state) => {
      const newMap = new Map(state.navHistory);
      newMap.set(code, history);
      return { navHistory: newMap };
    });

    return history;
  },

  getFundHoldings: async (code: string) => {
    const cached = get().holdingsCache.get(code);
    if (cached) return cached;

    const holdings = await apiGetFundHoldings(code);

    set((state) => {
      const newMap = new Map(state.holdingsCache);
      newMap.set(code, holdings);
      return { holdingsCache: newMap };
    });

    return holdings;
  },

  clearError: () => {
    set({ error: null });
  },

  clearSelection: () => {
    set({ selectedFund: null });
  },
}));

/**
 * Calculate volatility from NAV history
 */
function calculateVolatility(navHistory: FundNAV[]): number {
  if (navHistory.length < 2) return 15;

  const returns: number[] = [];
  for (let i = 1; i < navHistory.length; i++) {
    const prev = navHistory[i - 1];
    const curr = navHistory[i];
    const prevNav = prev?.accNav ?? prev?.nav ?? 0;
    const currNav = curr?.accNav ?? curr?.nav ?? 0;
    if (prevNav > 0) {
      returns.push((currNav - prevNav) / prevNav);
    }
  }

  if (returns.length === 0) return 15;

  const mean = returns.reduce((a, b) => a + b, 0) / returns.length;
  const variance = returns.reduce((sum, r) => sum + (r - mean) ** 2, 0) / returns.length;
  const stdDev = Math.sqrt(variance);

  // Annualized volatility (assuming 252 trading days)
  return Number((stdDev * Math.sqrt(252) * 100).toFixed(2));
}

/**
 * Calculate max drawdown from NAV history
 */
function calculateMaxDrawdown(navHistory: FundNAV[]): number {
  if (navHistory.length === 0) return 0;

  let maxDrawdown = 0;
  const first = navHistory[0];
  let peak = first?.accNav ?? first?.nav ?? 0;

  for (const point of navHistory) {
    const nav = point?.accNav ?? point?.nav ?? 0;
    if (nav > peak) {
      peak = nav;
    }
    const drawdown = (peak - nav) / peak;
    if (drawdown > maxDrawdown) {
      maxDrawdown = drawdown;
    }
  }

  return Number((-maxDrawdown * 100).toFixed(2));
}

/**
 * Calculate Sharpe ratio from NAV history
 */
function calculateSharpeRatio(navHistory: FundNAV[]): number {
  if (navHistory.length < 2) return 0.5;

  const returns: number[] = [];
  for (let i = 1; i < navHistory.length; i++) {
    const prev = navHistory[i - 1];
    const curr = navHistory[i];
    const prevNav = prev?.accNav ?? prev?.nav ?? 0;
    const currNav = curr?.accNav ?? curr?.nav ?? 0;
    if (prevNav > 0) {
      returns.push((currNav - prevNav) / prevNav);
    }
  }

  if (returns.length === 0) return 0.5;

  const mean = returns.reduce((a, b) => a + b, 0) / returns.length;
  const variance = returns.reduce((sum, r) => sum + (r - mean) ** 2, 0) / returns.length;
  const stdDev = Math.sqrt(variance);

  if (stdDev === 0) return 0;

  // Annualized Sharpe ratio (assuming 252 trading days, risk-free rate of 2%)
  const riskFreeRate = 0.02 / 252;
  return Number(((mean - riskFreeRate) / stdDev * Math.sqrt(252)).toFixed(2));
}

File: fund/lib/stores/fund_store.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../models/fund.dart';
import '../services/fund_api.dart';

final fundApiServiceProvider = Provider<FundApiService>((ref) => FundApiService());

final fundStoreProvider =
    StateNotifierProvider<FundStoreNotifier, FundState>((ref) {
  return FundStoreNotifier(ref.watch(fundApiServiceProvider));
});

class FundState {
  final List<FundInfo> searchResults;
  final FundInfo? selectedFund;
  final List<FundNAV> navHistory;
  final List<FundHolding> topHoldings;
  final bool isLoading;
  final String? error;

  FundState({
    this.searchResults = const [],
    this.selectedFund,
    this.navHistory = const [],
    this.topHoldings = const [],
    this.isLoading = false,
    this.error,
  });

  FundState copyWith({
    List<FundInfo>? searchResults,
    FundInfo? selectedFund,
    List<FundNAV>? navHistory,
    List<FundHolding>? topHoldings,
    bool? isLoading,
    String? error,
  }) =>
      FundState(
        searchResults: searchResults ?? this.searchResults,
        selectedFund: selectedFund ?? this.selectedFund,
        navHistory: navHistory ?? this.navHistory,
        topHoldings: topHoldings ?? this.topHoldings,
        isLoading: isLoading ?? this.isLoading,
        error: error,
      );
}

class FundStoreNotifier extends StateNotifier<FundState> {
  final FundApiService _api;

  FundStoreNotifier(this._api) : super(FundState());

  Future<void> searchFunds(String keyword) async {
    if (keyword.trim().isEmpty) {
      state = state.copyWith(searchResults: []);
      return;
    }

    state = state.copyWith(isLoading: true, error: null);
    try {
      final results = await _api.searchFunds(keyword);
      state = state.copyWith(searchResults: results, isLoading: false);
    } catch (e) {
      state = state.copyWith(error: e.toString(), isLoading: false);
    }
  }

  Future<void> selectFund(String code) async {
    state = state.copyWith(isLoading: true, error: null);
    try {
      final fundInfo = await _api.getFundInfo(code);
      if (fundInfo == null) {
        state = state.copyWith(
          error: '无法获取基金信息',
          isLoading: false,
        );
        return;
      }

      final navHistory = await _api.getFundNAV(code);
      final topHoldings = await _api.getFundHoldings(code);

      state = state.copyWith(
        selectedFund: fundInfo,
        navHistory: navHistory,
        topHoldings: topHoldings,
        isLoading: false,
      );
    } catch (e) {
      state = state.copyWith(error: e.toString(), isLoading: false);
    }
  }

  void clearSelection() {
    state = FundState();
  }

  void clearError() {
    state = state.copyWith(error: null);
  }

  // Calculate risk metrics from NAV history
  RiskMetrics calculateRiskMetrics() {
    if (state.navHistory.length < 2) {
      return RiskMetrics(
        volatility: 0,
        maxDrawdown: 0,
        sharpeRatio: 0,
        correlationToMarket: 0,
      );
    }

    final returns = <double>[];
    for (var i = 1; i < state.navHistory.length; i++) {
      final prev = state.navHistory[i - 1].nav;
      final curr = state.navHistory[i].nav;
      if (prev != 0) {
        returns.add((curr - prev) / prev);
      }
    }

    if (returns.isEmpty) {
      return RiskMetrics(
        volatility: 0,
        maxDrawdown: 0,
        sharpeRatio: 0,
        correlationToMarket: 0,
      );
    }

    // Calculate volatility (standard deviation of returns)
    final mean = returns.reduce((a, b) => a + b) / returns.length;
    final variance =
        returns.map((r) => (r - mean) * (r - mean)).reduce((a, b) => a + b) /
            returns.length;
    final volatility = variance > 0 ? variance * 100 : 0;

    // Calculate max drawdown
    double maxDrawdown = 0;
    double peak = state.navHistory.first.nav;
    for (final nav in state.navHistory) {
      if (nav.nav > peak) peak = nav.nav;
      final drawdown = (peak - nav.nav) / peak;
      if (drawdown > maxDrawdown) maxDrawdown = drawdown;
    }

    // Simplified Sharpe Ratio (assuming 0% risk-free rate)
    final sharpeRatio = mean != 0 ? (mean / (variance > 0 ? variance : 0.001)) : 0;

    return RiskMetrics(
      volatility: volatility * 100,
      maxDrawdown: maxDrawdown * 100,
      sharpeRatio: sharpeRatio * 100,
      correlationToMarket: 0.5, // Simplified - would need market data for real value
    );
  }

  // Calculate rating based on risk metrics (1-5 stars)
  int calculateRating(RiskMetrics metrics) {
    var score = 0;

    // Lower volatility is better
    if (metrics.volatility < 10) score += 2;
    else if (metrics.volatility < 20) score += 1;

    // Lower max drawdown is better
    if (metrics.maxDrawdown < 10) score += 2;
    else if (metrics.maxDrawdown < 20) score += 1;

    // Higher Sharpe ratio is better
    if (metrics.sharpeRatio > 1) score += 1;

    return (score + 1).clamp(1, 5);
  }
}
